File: Online-judge/leetcode/sort-an-array.py
from typing import List

# TL NOT solved 

class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        if len(nums) > 1:
            mid = len(nums) // 2
            leftList = self.sortArray(nums[:mid])
            rightList = self.sortArray(nums[mid:])
            return self.merge(leftList, rightList)
        else:
            return nums

    def merge(self, leftList: List[int], rightList: List[int]) -> List[int]:
        i, j = 0, 0
        newList: List[int] = []
        while i < len(leftList) and j < len(rightList):
            if leftList[i] < rightList[j]:
                newList.append(leftList[i])
                i += 1
            else:
                newList.append(rightList[j])
                j += 1
        while i < len(leftList):
            newList.append(leftList[i])
            i += 1
        while j < len(rightList):
            newList.append(rightList[j])
        return newList 

File: Online-judge/leetcode/sort-an-array.cpp
# include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/sort-an-array/description/
 * Time complexity O(NlogN) sort taks O(NlogN)
 * Space complexity O(1)
*/

class Solution {
private:

	void swap(int *a, int *b) {
		int temp = *a;
		*a = *b;
		*b = temp;
	}

	int partition(vector<int> &nums, int low, int high) {
		int counter = low-1;
		int pivot = nums[high];
		for(int i = low; i < high; i++) {
			if (nums[i] <= pivot) {
				counter++;
				swap(&nums[counter], &nums[i]);
			}
		}
		swap(&nums[counter+1], &nums[high]);
		return counter+1;
	}

	void QuickSort(vector<int>& nums, int low, int high) {
		if(low < high) {
			int pi = partition(nums, low, high);
			QuickSort(nums, low, pi-1);
			QuickSort(nums, pi+1, high);
		}
	}

public:
    vector<int> sortArray(vector<int>& nums) {
    	if(nums.size() < 2) return nums;
    	QuickSort(nums, 0, nums.size()-1);
    	return nums;
    }
};